packages/next/build/swc/src/next_dynamic.rs

```diff
@@ -0,0 +1,203 @@
+use swc_atoms::js_word;
+use swc_common::{FileName, DUMMY_SP};
+use swc_ecmascript::ast::{
+  ArrayLit, ArrowExpr, BinExpr, BinaryOp, BlockStmtOrExpr, CallExpr, Expr, ExprOrSpread,
+  ExprOrSuper, Ident, ImportDecl, ImportSpecifier, KeyValueProp, Lit, MemberExpr, ObjectLit, Prop,
+  PropName, PropOrSpread, Str, StrKind,
+};
+use swc_ecmascript::utils::{
+  ident::{Id, IdentLike},
+  HANDLER,
+};
+use swc_ecmascript::visit::{Fold, FoldWith};
+
+pub fn next_dynamic(filename: FileName) -> impl Fold {
+  NextDynamicPatcher {
+    filename,
+    dynamic_bindings: vec![],
+  }
+}
+
+#[derive(Debug)]
+struct NextDynamicPatcher {
+  filename: FileName,
+  dynamic_bindings: Vec<Id>,
+}
+
+impl Fold for NextDynamicPatcher {
+  fn fold_import_decl(&mut self, decl: ImportDecl) -> ImportDecl {
+    let ImportDecl {
+      ref src,
+      ref specifiers,
+      ..
+    } = decl;
+    if &src.value == "next/dynamic" {
+      for specifier in specifiers {
+        if let ImportSpecifier::Default(default_specifier) = specifier {
+          self.dynamic_bindings.push(default_specifier.local.to_id());
+        }
+      }
+    }
+
+    decl
+  }
+
+  fn fold_call_expr(&mut self, expr: CallExpr) -> CallExpr {
+    let mut expr = expr.fold_children_with(self);
+    if let ExprOrSuper::Expr(i) = &expr.callee {
+      if let Expr::Ident(identifier) = &**i {
+        if self.dynamic_bindings.contains(&identifier.to_id()) {
+          if expr.args.len() == 0 {
+            HANDLER.with(|handler| {
+              handler
+                .struct_span_err(
+                  identifier.span,
+                  "next/dynamic requires at least one argument",
+                )
+                .emit()
+            });
+          } else if expr.args.len() > 2 {
+            HANDLER.with(|handler| {
+              handler
+                .struct_span_err(identifier.span, "next/dynamic only accepts 2 arguments")
+                .emit()
+            });
+          }
+
+          let mut import_specifier = None;
+          if let Expr::Arrow(ArrowExpr {
+            body: BlockStmtOrExpr::Expr(e),
+            ..
+          }) = &*expr.args[0].expr
+          {
+            if let Expr::Call(CallExpr { args: a, .. }) = &**e {
+              if let Expr::Lit(Lit::Str(Str { value, .. })) = &*a[0].expr {
```

Comment:
Indexing error for `a[0]` will not contain useful information.

```ts
import l from 'next/dynamic';

const P = l(() => foo("str"));
const P = l(() => foo()); // This will panic

```